### PR TITLE
Speed up `EntrySortIterator` by 20%.

### PR DESCRIPTION
**Pull request by jeschkies**
**What this PR does / why we need it**:
This patch replace the underlying heap of the `EntrySortIterator`. Instead of using a min or max heap the entry iterators are sorted at the beginning. Once the iteration starts only the first entry iterator is re-positioned depending on its current value. This will yield the same runtime complexity as the current heap based solution. However, since the new algorithm accesses the first element it can be cached. Thus the number of access calls to `Entry().Timestamp` are halved.

```
› benchstat before.txt after.txt
name                        old time/op  new time/op  delta
SortIterator/merge_sort-16  4.40ms ± 4%  4.38ms ± 3%     ~     (p=0.400 n=9+10)
SortIterator/sort-16        3.65ms ± 2%  2.80ms ± 3%  -23.25%  (p=0.000 n=9+10)
```

**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Special notes for your reviewer**:

**Checklist**
- [ ] Documentation added
- [ ] Tests updated
- [ ] Add an entry in the `CHANGELOG.md` about the changes.

### CODE REVIEW

File: pkg/iter/entry_iterator.go [hunk 1]
```diff
@@ -344,28 +342,61 @@ func NewSortEntryIterator(is []EntryIterator, direction logproto.Direction) Entr
...
+	case un1 < un2:
+		return true
+	case un1 > un2:
+		return false
+	default: // un1 == un2:
+		return it.is[i].Labels() < it.is[j].Labels()
+	}
+}
+
+func (it *entrySortIterator) lessThan(t1 time.Time, l1 string, j int) bool {
```
Comment (jeschkies) [inline pkg/iter/entry_iterator.go:370]: We probably should bass `un2` instead.

File: pkg/iter/entry_iterator.go [hunk 2]
```diff
@@ -344,28 +342,61 @@ func NewSortEntryIterator(is []EntryIterator, direction logproto.Direction) Entr
...
+	t2 := it.is[j].Entry().Timestamp
+	if it.dir == logproto.BACKWARD {
+		t2, t1 = t1, t2
+	}
```
Comment (jeschkies) [inline pkg/iter/entry_iterator.go:374]: This is nasty. Ideally we would define the `lessThan` function once at the beginning.

File: pkg/iter/entry_iterator.go [hunk 3]
```diff
@@ -374,35 +405,55 @@ func (i *entrySortIterator) init() {
...
+	i.is = tmp
+	sort.Slice(i.is, i.less)
+}
 
-	// We can now clear the list of input iterators to merge, given they have all
-	// been processed and the non empty ones have been pushed to the heap
-	i.is = nil
+func (i *entrySortIterator) fix() {
+	t1 := i.is[0].Entry().Timestamp
+	l1 := i.is[0].Labels()
```
Comment (jeschkies) [inline pkg/iter/entry_iterator.go:414]: This yields the 20% speedup.

File: pkg/iter/entry_iterator.go [hunk 4]
```diff
@@ -374,35 +405,55 @@ func (i *entrySortIterator) init() {
...
+	// First element is out of place. So we reposition it.
+	index := sort.Search(len(i.is), func(in int) bool { return i.lessThan(t1, l1, in) })
+
+	head := i.is[0]
+	if index == len(i.is) {
+		i.is = append(i.is[1:], head)
+	} else {
+		i.is = append(i.is[1:index+1], i.is[index:]...)
+		i.is[index] = head
+	}
```
Comment (jeschkies) [inline pkg/iter/entry_iterator.go:430]: This has the same runtime as `heap.Fix()`.

File: pkg/iter/entry_iterator.go [hunk 5]
```diff
@@ -426,11 +481,6 @@ func (i *entrySortIterator) Error() error {
 }
 
 func (i *entrySortIterator) Close() error {
-	for i.heap.Len() > 0 {
```
Comment (cyriltovena) [inline pkg/iter/entry_iterator.go:429]: Still need to close if iteration is not finished.

File: vendor/modules.txt [hunk 6]
```diff
@@ -1028,6 +1028,9 @@ github.com/prometheus/prometheus/util/teststorage
 github.com/prometheus/prometheus/util/testutil
 github.com/prometheus/prometheus/util/treecache
 github.com/prometheus/prometheus/web/api/v1
+# github.com/psilva261/timsort/v2 v2.0.0
+## explicit; go 1.13
+github.com/psilva261/timsort/v2
```
Comment (cyriltovena) [inline vendor/modules.txt:1033]:
```suggestion
```

File: pkg/iter/entry_iterator.go [hunk 7]
```diff
@@ -374,35 +405,55 @@ func (i *entrySortIterator) init() {
...
+
+	// shortcut
+	if len(i.is) <= 1 || i.lessThan(t1, l1, 1) {
+		return
+	}
+
+	// First element is out of place. So we reposition it.
+	index := sort.Search(len(i.is), func(in int) bool { return i.lessThan(t1, l1, in) })
+
+	head := i.is[0]
```
Comment (cyriltovena) [inline pkg/iter/entry_iterator.go:424]: head could be defined at the top and reused everywhere.

File: pkg/iter/entry_iterator.go [hunk 8]
```diff
@@ -344,28 +342,61 @@ func NewSortEntryIterator(is []EntryIterator, direction logproto.Direction) Entr
...
-	}
+
+	result := &entrySortIterator{is: is, dir: direction}
 	return result
 }
 
-// init initialize the underlaying heap
+func (it *entrySortIterator) less(i, j int) bool {
+	t1, t2 := it.is[i].Entry().Timestamp, it.is[j].Entry().Timestamp
+	if it.dir == logproto.BACKWARD {
```
Comment (cyriltovena) [inline pkg/iter/entry_iterator.go:352]: You'll need to carefully merge https://github.com/grafana/loki/pull/5289.

**Review by cyriltovena — COMMENTED** [review]
Looking good please make sure you correctly merge freshly change there.

File: pkg/iter/entry_iterator.go (hunk 7 above)
Comment (jeschkies) [reply pkg/iter/entry_iterator.go:396]: Nice that gave the speed up to 25%.

File: pkg/iter/entry_iterator.go [hunk 9]
```diff
@@ -347,36 +376,58 @@ func (i *entrySortIterator) init() {
...
+	l1 := head.Labels()
+	h1 := head.StreamHash()
+
+	// shortcut
+	if len(i.is) <= 1 || i.lessByValue(t1, l1, h1, 1) {
+		return
+	}
+
+	// First element is out of place. So we reposition it.
+	i.is = i.is[1:] // drop head
```
Comment (jeschkies) [inline pkg/iter/entry_iterator.go:395]: I'm sure we can factor this into line 401 but it gave me too much of a headache.

File: pkg/iter/entry_iterator.go [hunk 10]
```diff
@@ -404,8 +455,8 @@ func (i *entrySortIterator) Error() error {
 }
 
 func (i *entrySortIterator) Close() error {
-	for i.heap.Len() > 0 {
-		if err := i.heap.Pop().(EntryIterator).Close(); err != nil {
+	if len(i.is) > 0 {
+		if err := i.is[0].Close(); err != nil {
```
Comment (cyriltovena) [inline pkg/iter/entry_iterator.go:459]: That doesn't close all remaining iterator like before.

Comment (jeschkies) [reply pkg/iter/entry_iterator.go:459]: I'm not sure I understand the original code then. It just closes the head of the heap doesn't it?

Comment (jeschkies) [reply pkg/iter/entry_iterator.go:459]: Sorry. Misread `for` as an `if`.

File: pkg/iter/entry_iterator.go [hunk 11]
```diff
@@ -320,25 +318,56 @@ func NewSortEntryIterator(is []EntryIterator, direction logproto.Direction) Entr
 	result := &entrySortIterator{is: is}
 	switch direction {
 	case logproto.BACKWARD:
-		result.heap = &iteratorSortHeap{iteratorHeap: make([]EntryIterator, 0, len(is)), byAscendingTime: false, byAlphabetical: true}
+		result.byAscendingTime = false
+		result.byAlphabetical = true
```
Comment (MasslessParticle) [inline pkg/iter/entry_iterator.go:322]: Is there any time we don't want `byAlphabetical` to be true? Could this just go in the initialization?

File: pkg/iter/entry_iterator.go [hunk 12]
```diff
@@ -320,25 +318,56 @@ func NewSortEntryIterator(is []EntryIterator, direction logproto.Direction) Entr
...
 		panic("bad direction")
 	}
 	return result
 }
 
-// init initialize the underlaying heap
+func (i *entrySortIterator) lessByIndex(k, j int) bool {
+	t1, t2 := i.is[k].Entry().Timestamp.UnixNano(), i.is[j].Entry().Timestamp.UnixNano()
+	if t1 == t2 {
+		if i.byAlphabetical {
```
Comment (MasslessParticle) [inline pkg/iter/entry_iterator.go:335]: Same here. If `byAlphabetical` is always true, we don't need to care about the other case

File: pkg/iter/entry_iterator.go (hunk 11 above)
Comment (jeschkies) [reply pkg/iter/entry_iterator.go:322]: The `main` version has the same logic. @cyriltovena when would `byAlphabetical` be false?

Comment (cyriltovena) [reply pkg/iter/entry_iterator.go:322]: The heap was shared with the Merge iterator which is not anymore so I think we can get rid of it.

The merge one is also always going to be non alphabetical I believe.

Comment (jeschkies) [reply pkg/iter/entry_iterator.go]: Done.

**Review by cyriltovena — APPROVED** [review]
LGTM

**Pull request merged**